### PR TITLE
Port env mods tests to Windows

### PR DESCRIPTION
**Pull request by nicolecheetham**

### CODE REVIEW

File: lib/spack/spack/test/data/sourceme_second.bat [hunk 1]
```diff
@@ -0,0 +1,10 @@
+@echo off
+rem "C:\lib\spack\spack\test\data
+rem
+rem Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+rem Spack Project Developers. See the top-level COPYRIGHT file for details.
+rem
+rem SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+set PATH_LIST=\path\first;\path\second;\path\fourth
```
Comment (johnwparent) [inline lib/spack/spack/test/data/sourceme_second.bat]: I think the more direct parallel here would be absolute Windows style paths (i.e. `C:\path\first` or whatever drive letter you want)

File: lib/spack/spack/test/environment_modifications.py [hunk 2]
```diff
@@ -62,18 +63,28 @@ def test_exclude_paths_from_inspection():
     assert len(env) == 0
 
 
+def begin_esc():
+    return "\\" if sys.platform == "win32" else "/"
+
+
+def shell_file():
```
Comment (johnwparent) [inline lib/spack/spack/test/environment_modifications.py]: (minor nits)
A. Can we rename this to `shell_extension` as that's what it's actually returning
B. Can this be a module attribute so we're not constantly invoking a method.

File: lib/spack/spack/util/environment.py [hunk 3]
```diff
@@ -767,7 +767,7 @@ def from_sourcing_file(
 
         # Compute the environments before and after sourcing
         before = sanitize(
-            environment_after_sourcing_files(os.devnull, **kwargs),
+            environment_after_sourcing_files("cd %CD%", **kwargs),
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: This should retain its cross platform compatibility (the current approach does not work on Linux/MacOS). Perhaps abstracting behind a function or module attribute would be ideal.

For context for other reviewers:
`os.devnul` isn't a value that's meaningful in this context on Windows, so we must affect some other no-op operation here. Changing directory to the current directory is an innocuous substitute.

File: lib/spack/spack/util/executable.py [hunk 4]
```diff
@@ -226,15 +226,16 @@ def streamify(arg, mode):
...
+        cmds = self.exe + list(args)
+        cmd = ['"' + c + '"' if "import" in c else c for c in cmds]
 
         escaped_cmd = ["'%s'" % arg.replace("'", "'\"'\"'") for arg in cmd]
         cmd_line_string = " ".join(escaped_cmd)
         tty.debug(cmd_line_string)
 
         try:
             proc = subprocess.Popen(
-                cmd, stdin=istream, stderr=estream, stdout=ostream, env=env, close_fds=False
```
Comment (johnwparent) [inline lib/spack/spack/util/executable.py]: I think this is going to cause issues, but I need to think a bit more about how we can rework this for this use case and to preserve previous behavior.

File: lib/spack/spack/test/environment_modifications.py [hunk 5]
```diff
@@ -62,18 +63,28 @@ def test_exclude_paths_from_inspection():
     assert len(env) == 0
 
 
+def begin_esc():
```
Comment (johnwparent) [inline lib/spack/spack/test/environment_modifications.py]: I think the usage of this method in this test module is not properly testing the expected behavior of the test module. On Linux/Mac the `/` character is the root of the filesystem in addition to the path separator, however on Windows those roles are the drive component and the backslash. The previous linux behavior uses the `/` because its testing absolute paths as members of path environment variables, so the Windows conversion should mirror that behavior.

**Comment by johnwparent** [comment]
Left an initial review, some small nits, some larger changes that will need to be made before CI passes and this can be merged.

**Comment by johnwparent** [comment]
@spackbot fix style

File: lib/spack/spack/util/environment.py [hunk 6]
```diff
@@ -767,9 +767,7 @@ def from_sourcing_file(
 
         # Compute the environments before and after sourcing
         before = sanitize(
-            environment_after_sourcing_files(os.devnull, **kwargs),
-            exclude=exclude,
-            include=include,
+            environment_after_sourcing_files("cd %CD%", **kwargs), exclude=exclude, include=include
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: This would only ever work on Windows and will break cross platform support. We should replace this with a variable that contains a cross platform solution.

**Review by johnwparent — CHANGES_REQUESTED** [review]

**Comment by nicolecheetham** [comment]
@spackbot fix style

**Comment by nicolecheetham** [comment]
@spackbot fix style

File: lib/spack/spack/util/environment.py [hunk 7]
```diff
@@ -807,7 +806,7 @@ def from_environment_diff(
         modified_variables.sort()
 
         def return_separator_if_any(*args):
-            separators = ":", ";"
+            separators = ";" if sys.platform == "win32" else ":"
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: `os.pathsep` is what this logic is looking to achieve.

**Review by johnwparent — CHANGES_REQUESTED** [review]

File: lib/spack/spack/util/environment.py [hunk 8]
```diff
@@ -787,11 +786,9 @@ def from_environment_diff(
...
 
         def return_separator_if_any(*args):
-            separators = ":", ";"
-            for separator in separators:
-                for arg in args:
-                    if separator in arg:
-                        return separator
+            for arg in args:
+                if os.pathsep in arg:
+                    return os.pathsep
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: From the diff it looks like Unix supports both `;` and `:` separators so `os.pathsep` is not sufficient on that platform. 
I think it should be something like:
```suggestion
            separators = [os.pathsep] if IS_WINDOWS else [":", ";"]
            for separator in separators:
                for arg in args:
                    if separator in arg:
                        return separator
```

File: lib/spack/spack/util/executable.py (hunk 4 above)
Comment (nicolecheetham) [reply lib/spack/spack/util/executable.py]: I decided to keep the list comprehensions for now as I thought its more succinct

File: lib/spack/spack/util/executable.py [hunk 9]
```diff
@@ -233,6 +233,17 @@ def streamify(arg, mode):
         cmd_line_string = " ".join(escaped_cmd)
         tty.debug(cmd_line_string)
 
+        if self.exe[0].lower().startswith(("pwsh", "powershell", "cmd")):
+            spaced_args = [arg for arg in cmd if re.search(r'.*\s.*', arg)]
```
Comment (johnwparent) [inline lib/spack/spack/util/executable.py]: I think this check can be the same as the check below i.e. `if " " in arg`.

File: lib/spack/spack/util/environment.py [hunk 10]
```diff
@@ -745,11 +745,10 @@ def from_sourcing_file(
             ]
         )
 
+        no_op = "cd %CD%" if sys.platform == "win32" else os.devnull
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: I think we need a comment here as well, an actual shell command on Windows or a null FD on Unix is a bit of disparate choice. 

(nit pick) Can we also rename the variable to something like `null_fd` or similar? That way in addition to the comment the variable name conveys what it's supposed to achieve.

File: lib/spack/spack/util/executable.py [hunk 11]
```diff
@@ -233,6 +233,17 @@ def streamify(arg, mode):
         cmd_line_string = " ".join(escaped_cmd)
         tty.debug(cmd_line_string)
 
+        if self.exe[0].lower().startswith(("pwsh", "powershell", "cmd")):
+            spaced_args = [arg for arg in cmd if " " in arg]
+            cmd = " ".join([f'"{arg}"' if " " in arg else arg for arg in cmd])
+            if spaced_args:
+                tty.warn(
+                    "Spaces in command arguments can can confuse script parsing.",
```
Comment (johnwparent) [inline lib/spack/spack/util/executable.py]: Awesome update, thanks! 

Can we add some comments describing why and when this is required and can we make this warning more specific to Windows/the shells as we have a similar warning above that could be easily confused with this one.

File: lib/spack/spack/util/environment.py [hunk 12]
```diff
@@ -1048,11 +1048,12 @@ def _source_single_file(file_and_args, environment):
             [source_file, suppress_output, concatenate_on_success, dump_environment_cmd]
         )
 
+        # List fails to preserve nested double quotes on Windows
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: Lists do not modify their contents implicitly, the quote stripping is due to CMD's handling of quotes and the process of passing these strings from python to the system call. Please update the comment to something like "Passing this expression to CMD requires the use of Popen's string based args to avoid unexpected parsing by the underlying subprocess module https://docs.python.org/3.8/library/subprocess.html#notes"

File: lib/spack/spack/util/environment.py [hunk 13]
```diff
@@ -744,11 +744,11 @@ def from_sourcing_file(
             ]
         )
 
+        # Implements a no op during forwards operation on respective os
+        null_fd = "cd %CD%" if sys.platform == "win32" else os.devnull
```
Comment (johnwparent) [inline lib/spack/spack/util/environment.py]: Can we change this to `echo > NUL`. It achieves the same thing as `cd %CD%` with the added benefit of being a bit more semantically in line with the intent of the operation here.

**Review by johnwparent — DISMISSED** [review]

File: lib/spack/spack/util/environment.py [hunk 14]
```diff
@@ -1048,11 +1048,12 @@ def _source_single_file(file_and_args, environment):
             [source_file, suppress_output, concatenate_on_success, dump_environment_cmd]
         )
 
+        # Windows requires Popen's string arg to avoid unexpected parsing by subprocess
```
Comment (scheibelp) [inline lib/spack/spack/util/environment.py]: I'm not sure I follow:

* I see the docs for popen say that the args are assembled into a string on windows: https://docs.python.org/3/library/subprocess.html#subprocess.Popen
* The rules are https://docs.python.org/3/library/subprocess.html#converting-an-argument-sequence-to-a-string-on-windows, and I guess the idea is that these rules produce undesirable results?

In any case, I think the parsing is expected, but we just don't want it, so I think the wording should change.

File: lib/spack/spack/util/environment.py [hunk 15]
```diff
@@ -744,11 +744,11 @@ def from_sourcing_file(
             ]
         )
 
+        # Implements a no op during forwards operation on respective os
+        null_fd = "echo > NUL" if sys.platform == "win32" else os.devnull
```
Comment (scheibelp) [inline lib/spack/spack/util/environment.py]: I'm not sure how this fits the API of `environment_after_sourcing_files`: for that function the first argument is supposed to be a file path or a tuple. I can sort of see how this might work by tracing through the logic, but I think it might match the expected API better to pass it an empty file (e.g. created by `tempfile`).

File: lib/spack/spack/test/environment_modifications.py [hunk 16]
```diff
@@ -158,75 +181,155 @@ def test_unset(env):
...
+                ["C:\\dev\\spack_window"]
+                if sys.platform == "win32"
+                else [
+                    "C:\\",
+                    "C:\\Program Files",
+                    "C:\\Program Files (x86)",
+                    "C:\\Users",
+                    "C:\\ProgramData",
+                    "C:\\dev\\spack_window",
+                ]
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: I think it would be best to exclude the windows parameterizations on non-Windows (and the other way around too).

File: lib/spack/spack/test/environment_modifications.py [hunk 17]
```diff
@@ -158,75 +181,155 @@ def test_unset(env):
...
+            ),
+        ),
+        (
+            "EMPTY_PATH_LIST",
+            ["first", "middle", "last"],
+            make_pathlist([["path", "first"], ["path", "middle"], ["path", "last"]]),
+        ),
+        (
+            "NEWLY_CREATED_PATH_LIST",
+            ["first", "middle", "last"],
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: (question) could you call make_pathlist on this?

File: lib/spack/spack/test/environment_modifications.py [hunk 18]
```diff
@@ -62,16 +89,38 @@ def test_exclude_paths_from_inspection():
     assert len(env) == 0
 
 
+def make_pathlist(paths):
+    """Makes a fake list of platform specific paths"""
+    return os.pathsep.join(
+        [
+            driver + os.path.join(*path) if isinstance(path, list) else driver + path
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: It's slightly confusing to refer to this as driver on windows, so I'd prefer  to avoid that, and also to consolidate all uses of `driver` outside of this into something like a `make_path` function (and `make_pathlist` can call that on each element of a list).

**Review by scheibelp — CHANGES_REQUESTED** [review]
@nicolecheetham @johnwparent 

I have a few requests. Also can you merge/rebase with develop since this PR is somewhat old (apologies for taking so long to review)

File: lib/spack/spack/util/environment.py (hunk 14 above)
Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: I think what was intended is: "The parsing is expected, the behavior after parsing was not", I'll rephrase to indicate popen's strategy for parsing causes invalid command line invocation if specified in another way on Windows.

File: lib/spack/spack/test/environment_modifications.py (hunk 16 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Agreed

File: lib/spack/spack/test/environment_modifications.py (hunk 18 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Will do

File: lib/spack/spack/test/environment_modifications.py (hunk 17 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: No problem

File: lib/spack/spack/util/environment.py (hunk 15 above)
Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: `NUL` is the Windows equivalent of `/dev/null`. I need to remind myself why this doesn't work out of the box, will update when I have a better explanation. It may be that I read an old SO post that claimed `os.devnull` doesn't work on Windows.

Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: I retooled this, let me know what you think

File: lib/spack/spack/util/environment.py (hunk 14 above)
Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: Rephrased.

File: lib/spack/spack/test/environment_modifications.py (hunk 18 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Refactored.

File: lib/spack/spack/test/environment_modifications.py (hunk 16 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Done.

File: lib/spack/spack/test/environment_modifications.py (hunk 17 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Actually the way this is used by the test I think it makes sense to keep as is.

**Comment by johnwparent** [comment]
@scheibelp this is ready for another look at your convenience

File: lib/spack/spack/util/environment.py [hunk 19]
```diff
@@ -842,9 +842,19 @@ def from_sourcing_file(
...
         # Compute the environments before and after sourcing
+
+        # Windows cannot call nul (os.devnull), only pipe output to it
+        # so we call echo on our null fd and then pipe it to null
+        before_kwargs = {**kwargs}
+        if sys.platform == "win32":
+            if "source_command" not in kwargs:
+                before_kwargs["source_command"] = "echo"
+            if "suppress_output" not in kwargs:
+                before_kwargs["suppress_output"] = "> nul"
```
Comment (scheibelp) [inline lib/spack/spack/util/environment.py]: This default should be in  `environment_after_sourcing_files` itself, since I think it is generally applicable.

The `if` statement before this, `if "source_command" not in kwargs:` is the one doing something odd, so should probably stay here. I think the comment should move to that because that statement specifically is what it applies to (you can't `source /dev/null` on Windows).

File: lib/spack/spack/test/environment_modifications.py [hunk 20]
```diff
@@ -61,16 +87,40 @@ def test_exclude_paths_from_inspection():
...
+            ["a", "c"],
+            ["remove", "this"],
+            ["a", "d"],
+            ["duplicate"],
+            ["f", "g"],
+        ]
+    )
+    os.environ["PATH_LIST_WITH_SYSTEM_PATHS"] = os.environ["REMOVE_PATH_LIST"] + make_pathlist(
+        [["usr", "include"]]
+    )
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: This reorders `REMOVE_PATH_LIST` (comes first instead of last) - was that intended? One of the behaviors we check is the ability to move entries like `/usr/include` to the end

File: lib/spack/spack/test/environment_modifications.py [hunk 21]
```diff
@@ -157,75 +181,107 @@ def test_unset(env):
...
+                "C:\\Program Files (x86)",
+                "C:\\Users",
+                "C:\\ProgramData",
+                "C:\\dev\\spack_window",
+            ],
+            ["C:\\dev\\spack_window"],
+            marks=pytest.mark.skipif(sys.platform != "win32", reason="Skip Windows paths on *nix"),
+        ),
+        pytest.param(
+            miscellaneous_paths,
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: I think either: 

* this should be renamed to `miscellaneous_unix_paths`
* or, `miscellaneous_paths` should be renamed to `miscellaneous_system_paths` and should be a fixture that returns os-appropriate paths

It's a bit awkward that this does things two different ways (and the variable naming is IMO hard to follow).

File: lib/spack/spack/test/environment_modifications.py [hunk 22]
```diff
@@ -61,16 +87,40 @@ def test_exclude_paths_from_inspection():
     assert len(env) == 0
 
 
+def make_path(*path):
+    drive = "C:\\" if sys.platform == "win32" else "/"
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: A short comment would be useful like "given components like a, b, c, form a valid absolute path for the operating system being tested; it does not have to exist"

File: lib/spack/spack/test/environment_modifications.py [hunk 23]
```diff
@@ -157,75 +181,107 @@ def test_unset(env):
...
-    expected = "/path/first:/path/middle:/path/last"
-    assert os.environ["NEWLY_CREATED_PATH_LIST"] == expected
+    assert not os.environ["PATH_LIST_WITH_SYSTEM_PATHS"].startswith(
+        make_pathlist([["usr", "include" + os.pathsep]])
+    )
+    assert os.environ["PATH_LIST_WITH_SYSTEM_PATHS"].endswith(make_pathlist([["usr", "include"]]))
 
-    assert os.environ["REMOVE_PATH_LIST"] == "/a/b:/a/c:/a/d:/f/g"
-
-    assert not os.environ["PATH_LIST_WITH_SYSTEM_PATHS"].startswith("/usr/include:")
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py:226]: I think this check is still useful

Also, IMO it would make sense to move the unix-specific checks into their own test (and in general, tests that are imbuing paths like `/usr/include` with special meaning)

**Review by scheibelp — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/environment_modifications.py (hunk 20 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Not sure if it was intentional as I didn't write, but I can't see a justification for doing so, I'll restore to previous.

**Comment by johnwparent** [comment]
@scheibelp comments addressed, ready for another round.

File: lib/spack/spack/test/environment_modifications.py [hunk 24]
```diff
@@ -61,16 +87,44 @@ def test_exclude_paths_from_inspection():
...
+            ["duplicate"],
+            ["a", "c"],
+            ["remove", "this"],
+            ["a", "d"],
+            ["duplicate"],
+            ["f", "g"],
+        ]
+    )
+    os.environ["PATH_LIST_WITH_SYSTEM_PATHS"] = (
+        make_pathlist([["usr", "include"]]) + os.environ["REMOVE_PATH_LIST"]
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: I think this has to be `pathsep.join`-ed to avoid https://github.com/spack/spack/actions/runs/15862629572/job/44723326743?pr=44872

File: lib/spack/spack/test/environment_modifications.py [hunk 25]
```diff
@@ -158,75 +210,96 @@ def test_unset(env):
...
-    ]
-    assert filtered == expected
+    filtered = filter_system_paths(system_paths_for_os + non_system_paths_for_os)
+    assert filtered == non_system_paths_for_os
 
 
-# TODO 27021
-@pytest.mark.not_on_windows("Not supported on Windows (yet)")
-def test_set_path(env):
+@pytest.mark.parametrize("name,elements,separator", [("A", ["foo", "bar", "baz"], os.pathsep)])
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py]: This parameterization has a single instance - IMO there should be a comment (about what kinds of parameterizations would make sense to add) or this should not be parameterized.

File: lib/spack/spack/test/environment_modifications.py [hunk 26]
```diff
@@ -158,75 +210,96 @@ def test_unset(env):
...
-    env.prepend_path("EMPTY_PATH_LIST", "/path/first")
+    env.prepend_path(path_name, make_path("path", elements[0]))
+    env.append_path(path_name, make_path("path", elements[1]))
+    env.append_path(path_name, make_path("path", elements[2]))
 
-    env.append_path("NEWLY_CREATED_PATH_LIST", "/path/middle")
-    env.append_path("NEWLY_CREATED_PATH_LIST", "/path/last")
-    env.prepend_path("NEWLY_CREATED_PATH_LIST", "/path/first")
+    env.remove_path("REMOVE_PATH_LIST", make_path("remove", "this"))
+    env.remove_path("REMOVE_PATH_LIST", make_path("duplicate") + os.sep)
```
Comment (scheibelp) [inline lib/spack/spack/test/environment_modifications.py:238]: Can the parameterization be undone for this? Right now this test is a bit confusing because it looks like some parts are re-executed with no variation based on the parameters (technically not a problem, but a bit odd to be doing that).

File: lib/spack/spack/util/environment.py [hunk 27]
```diff
@@ -874,9 +874,17 @@ def from_sourcing_file(
             ]
         )
 
+        # Implements a no op during forwards operation on respective os
         # Compute the environments before and after sourcing
+
+        before_kwargs = {**kwargs}
+        if sys.platform == "win32":
+            # Windows cannot call nul (os.devnull), only pipe output to it
+            # so we call echo on our null fd and then pipe it to null
```
Comment (scheibelp) [inline lib/spack/spack/util/environment.py]: IMO the following is clearer

> Windows cannot source os.devnull, but it can echo it, so we override the "source" action in the method that extracts the env (environment_after_sourcing_files)

* "call" -> "source"
* "only pipe output to it": we are echoing `os.devnull`, so we are doing more than directing output to it

File: lib/spack/spack/util/environment.py [hunk 28]
```diff
@@ -874,9 +874,17 @@ def from_sourcing_file(
             ]
         )
 
+        # Implements a no op during forwards operation on respective os
```
Comment (scheibelp) [inline lib/spack/spack/util/environment.py]: This comment should go below (to the "before" part), although as it is I think something more like:

> First look at the environment after doing nothing to establish baseline

Might be clearer.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Some additional requests, mostly about docs and minor refactoring.

File: lib/spack/spack/test/environment_modifications.py (hunk 26 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py:238]: Done, unsure why this was parameterized in the first place.

File: lib/spack/spack/test/environment_modifications.py (hunk 25 above)
Comment (johnwparent) [reply lib/spack/spack/test/environment_modifications.py]: Could not find a reason this was parameterized nor could I think of more parameterizations, removed the parameterization.

File: lib/spack/spack/util/environment.py (hunk 28 above)
Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: Done!

File: lib/spack/spack/util/environment.py (hunk 27 above)
Comment (johnwparent) [reply lib/spack/spack/util/environment.py]: Thanks! Done!

**Comment by johnwparent** [comment]
@scheibelp ready for another pass when you are!

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
(close/reopen to deal with stalled docs rebuild)
